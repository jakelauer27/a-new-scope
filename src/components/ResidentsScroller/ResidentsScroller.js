import React, { Component } from 'react'
import propTypes from 'prop-types'
import { uid } from 'react-uid'

export default class ResidentsScroller extends Component {
  constructor(props) {
    super(props)
    this.state = {
      residents: [],
      content1: '',
      content2: '',
      currentIndex: 0
    }
  }

  componentDidMount() {
    const { content1, content2 } = this.props
    if (content1.length === 0) {
      this.setState({
        residents: [{ name: 'none', display: '' }]
      })
    } else if (typeof content1 !== 'string') {
      this.setState({
        residents: content1.map((resident, i) => {
          let formattedResident = { name: resident, display: 'none' }
          if (i === 0) {
            formattedResident = { name: resident, display: '' }
          }
          return formattedResident
        })
      })
    } else {
      this.setState({
        content1,
        content2
      })
    }
  }

  clickArrow = (direction) => {
    const { residents, currentIndex } = this.state
    let newIndex = direction + currentIndex
    if (newIndex < 0) { newIndex = residents.length - 1 }
    if (newIndex >= residents.length) { newIndex = 0 }

    const newResidents = residents.map((resident, i) => {
      let formattedResident
      if (i === newIndex) {
        formattedResident = { name: resident.name, display: '' }
      } else {
        formattedResident = { name: resident.name, display: 'none' }
      }
      return formattedResident
    })

    this.setState({
      currentIndex: newIndex,
      residents: newResidents
    })
  }

  render() {
    const { residents, content1, content2 } = this.state

    if (content1) {
      return (
        <div className="secInfo">
          <h1 className="homeworld">{content1}</h1>
          <h1 className="population">{content2}</h1>
        </div>
      )
    }

    return (
      <div className="residents-scroller">
        <i className="fas fa-caret-left resident-arrow" role="presentation" onClick={() => this.clickArrow(-1)} />
        <i className="fas fa-caret-right resident-arrow" role="presentation" onClick={() => this.clickArrow(1)} />
        {
          residents.map(resident => (<h4
            className={`resident ${resident.display}`}
            key={uid(resident)}
          >
            {resident.name}
          </h4>
          ))
        }
      </div>
    )
  }
}

ResidentsScroller.propTypes = {
  content1: propTypes.oneOfType([propTypes.string, propTypes.array]).isRequired,
  content2: propTypes.string
}

ResidentsScroller.defaultProps = {
  content2: ''
}
