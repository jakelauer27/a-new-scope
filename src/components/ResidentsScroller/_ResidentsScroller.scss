.residents-scroller {
  position: absolute;
  margin: 5px 15px;
  width: 70%;
  z-index: 2;
  bottom: 25px;
  
  .fa-caret-left,
  .fa-caret-right {
    bottom: -14px;
    color: white;
    padding: 6px 10px;
    position: absolute;
    transition: all .3s;
    z-index: 16;

    &:hover {
      cursor: pointer;
    }
  } 

  .fa-caret-left {
    left: -5px;
  }

  .fa-caret-right {
    right: -3px;
  }
  
  .resident {
    bottom: -25px;
    color: white;
    font-weight: 200;
    font-style: italic;
    text-align: center;
    position: relative;
  }  
  
  .none {
    display: none;
  }

}