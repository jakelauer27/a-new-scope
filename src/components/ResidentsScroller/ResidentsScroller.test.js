import React from 'react'
import { shallow } from 'enzyme'
import ResidentsScroller from './ResidentsScroller'

describe('residentsScroller', () => {
  let content1
  let content2
  let wrapper

  beforeEach(() => {
    content1 = ['Anikan', 'R2-D2']
    content2 = ''
    wrapper = shallow(<ResidentsScroller content1={content1} content2={content2} />)
  })

  it('Should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Should default to no residents', () => {
    wrapper = shallow(<ResidentsScroller
      content1={content1}
      content2={content2}
    />, { disableLifecycleMethods: true })
    const expectedState = {
      residents: [],
      content1: '',
      content2: '',
      currentIndex: 0
    }

    expect(wrapper.state()).toEqual(expectedState)
  })

  describe('ComponentDidMount', () => {
    it('Should set residents to array of formatted residents if there are multiple', () => {
      const expectedResidents = [
        { name: 'Anikan', display: '' },
        { name: 'R2-D2', display: 'none' }
      ]

      expect(wrapper.state().residents).toEqual(expectedResidents)
    })

    it('Should set residents if there are 0 residents', () => {
      content1 = []
      wrapper = shallow(<ResidentsScroller content1={content1} content2={content2} />)
      const expectedResidents = [{ name: 'none', display: '' }]

      expect(wrapper.state().residents).toEqual(expectedResidents)
    })

    it('Should set content1 and content2 if there is not a residents category', () => {
      content1 = 'Something Interesting'
      content2 = 'Something Else Interesting'
      wrapper = shallow(<ResidentsScroller content1={content1} content2={content2} />)
      const expectedState = {
        residents: [],
        content1: 'Something Interesting',
        content2: 'Something Else Interesting',
        currentIndex: 0
      }

      expect(wrapper.state()).toEqual(expectedState)
    })
  })

  describe('clickArrow', () => {
    it('Should increment the character index', () => {
      const expectedIndex = 1
      wrapper.find('.fa-caret-right').simulate('click')

      expect(wrapper.state().currentIndex).toBe(expectedIndex)
    })

    it('Should decrement the character index', () => {
      wrapper.state().currentIndex = 1
      const expectedIndex = 0
      wrapper.find('.fa-caret-left').simulate('click')

      expect(wrapper.state().currentIndex).toBe(expectedIndex)
    })

    it('Should reset the character index to 0 if incrementing past the length of the residents array', () => {
      wrapper.state().currentIndex = 1
      const expectedIndex = 0
      wrapper.find('.fa-caret-right').simulate('click')

      expect(wrapper.state().currentIndex).toBe(expectedIndex)
    })

    it('Should reset the character index to the length of the residents array if decrementing past 0', () => {
      wrapper.state().currentIndex = 0
      const expectedIndex = 1
      wrapper.find('.fa-caret-right').simulate('click')

      expect(wrapper.state().currentIndex).toBe(expectedIndex)
    })
  })
})
