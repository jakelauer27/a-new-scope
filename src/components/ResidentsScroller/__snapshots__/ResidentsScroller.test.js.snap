// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`residentsScroller Should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ResidentsScroller
    content1={
      Array [
        "Anikan",
        "R2-D2",
      ]
    }
    content2=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <i
          className="fas fa-caret-left resident-arrow"
          onClick={[Function]}
        />,
        <i
          className="fas fa-caret-right resident-arrow"
          onClick={[Function]}
        />,
        Array [
          <h4
            className="resident "
          >
            Anikan
          </h4>,
          <h4
            className="resident none"
          >
            R2-D2
          </h4>,
        ],
      ],
      "className": "residents-scroller",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fas fa-caret-left resident-arrow",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fas fa-caret-right resident-arrow",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      Object {
        "instance": null,
        "key": "uid1",
        "nodeType": "host",
        "props": Object {
          "children": "Anikan",
          "className": "resident ",
        },
        "ref": null,
        "rendered": "Anikan",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": "uid2",
        "nodeType": "host",
        "props": Object {
          "children": "R2-D2",
          "className": "resident none",
        },
        "ref": null,
        "rendered": "R2-D2",
        "type": "h4",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <i
            className="fas fa-caret-left resident-arrow"
            onClick={[Function]}
          />,
          <i
            className="fas fa-caret-right resident-arrow"
            onClick={[Function]}
          />,
          Array [
            <h4
              className="resident "
            >
              Anikan
            </h4>,
            <h4
              className="resident none"
            >
              R2-D2
            </h4>,
          ],
        ],
        "className": "residents-scroller",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fas fa-caret-left resident-arrow",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fas fa-caret-right resident-arrow",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        Object {
          "instance": null,
          "key": "uid1",
          "nodeType": "host",
          "props": Object {
            "children": "Anikan",
            "className": "resident ",
          },
          "ref": null,
          "rendered": "Anikan",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": "uid2",
          "nodeType": "host",
          "props": Object {
            "children": "R2-D2",
            "className": "resident none",
          },
          "ref": null,
          "rendered": "R2-D2",
          "type": "h4",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`residentsScroller Should match the snapshot if no content 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ResidentsScroller
    content1={
      Array [
        "Anikan",
        "R2-D2",
      ]
    }
    content2=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <i
          className="fas fa-caret-left resident-arrow"
          onClick={[Function]}
        />,
        <i
          className="fas fa-caret-right resident-arrow"
          onClick={[Function]}
        />,
        Array [],
      ],
      "className": "residents-scroller",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fas fa-caret-left resident-arrow",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fas fa-caret-right resident-arrow",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <i
            className="fas fa-caret-left resident-arrow"
            onClick={[Function]}
          />,
          <i
            className="fas fa-caret-right resident-arrow"
            onClick={[Function]}
          />,
          Array [],
        ],
        "className": "residents-scroller",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fas fa-caret-left resident-arrow",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fas fa-caret-right resident-arrow",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
